cli/crates/federated-dev/src/dev/websockets.rs

```diff
@@ -0,0 +1,234 @@
+use std::collections::HashMap;
+
+use ::axum::extract::ws::{self, WebSocket};
+use futures_util::{pin_mut, stream::SplitStream, SinkExt, StreamExt};
+use serde_json::json;
+use tokio::sync::mpsc;
+
+use self::messages::{Event, Message};
+
+use super::bus::GatewayWatcher;
+
+mod axum;
+mod messages;
+
+pub use axum::WebsocketService;
+
+pub type WebsocketSender = tokio::sync::mpsc::Sender<WebSocket>;
+pub type WebsocketReceiver = tokio::sync::mpsc::Receiver<WebSocket>;
+
+/// An actor that manages websocket connections for federated dev
+pub(crate) struct WebsocketAccepter {
+    sockets: WebsocketReceiver,
+    gateway: GatewayWatcher,
+}
+
+impl WebsocketAccepter {
+    pub fn new(sockets: WebsocketReceiver, gateway: GatewayWatcher) -> Self {
+        Self { sockets, gateway }
+    }
+
+    pub async fn handler(mut self) {
+        while let Some(connection) = self.sockets.recv().await {
+            let gateway = self.gateway.clone();
+
+            tokio::spawn(async move {
+                let Some((websocket, _init_payload)) = accept_websocket(connection).await else {
+                    log::warn!("Failed to accept websocket connection");
+                    return;
```

Comment:
Shouldn't an error be returned?

Reply:
Returned where?  The only way to get to this point at the moment is if the connection is closed.

Reply:
If you get an `Event::Subscribe` event for example, the `accept_websocket` function will return `None`.  It's an edge case though obviously

Reply:
If it gets an `Event::Subscribe` it will ignore it and continue looping.  It only returns `None` if the websocket stream ends - which means there's nowhere to return errors.

I do need to add timeouts on this accept call but I'm going to do that in [another issue](https://linear.app/grafbase/issue/GB-5792/handle-ws-accept-failures-in-the-cli)

Reply:
ah my bad, didn't notice the `while`.